contracts/oracle/BackedOracle.sol

```diff
@@ -0,0 +1,201 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+pragma solidity 0.8.9;
+
+import "./BackedOracleInterface.sol";
+
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+
+contract BackedOracle is AccessControlUpgradeable, AggregatorV2V3Interface {
+    uint8 public constant VERSION = 1;
+    uint8 public constant MAX_PERCENT_DIFFERENCE = 10;
+    uint32 public constant MAX_TIMESTAMP_AGE = 5 minutes;
+    uint32 public constant MIN_UPDATE_INTERVAL = 1 hours;
+
+    bytes32 public constant UPDATER_ROLE = keccak256("UPDATER_ROLE");
+
+    struct RoundData {
+        int192 answer;
+        uint32 timestamp;
+    }
+
+    uint8 private _decimals;
+    string private _description;
+
+    mapping(uint256 => RoundData) private _roundData;
+    uint80 private _latestRoundNumber;
+
+    constructor() {
+        initialize(0, "Backed Oracle Implementation", address(0), address(0));
+    }
+
+    function initialize(
+        uint8 __decimals,
+        string memory __description,
+
+        address __admin,
+        address __updater
+    ) public initializer {
+        _decimals = __decimals;
+        _description = __description;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, __admin);
+        _grantRole(UPDATER_ROLE, __updater);
+    }
+
+    function version() external view override returns (uint256) {
+        return VERSION;
+    }
+
+    function decimals() external view override returns (uint8) {
+        return _decimals;
+    }
+
+    function description() external view override returns (string memory) {
+        return _description;
+    }
+
+    function latestAnswer() external view override returns (int256) {
+        require(_latestRoundNumber != 0, "No data present");
+
+        return _roundData[_latestRoundNumber].answer;
+    }
+
+    function latestTimestamp() external view override returns (uint256) {
+        require(_latestRoundNumber != 0, "No data present");
+
+        return _roundData[_latestRoundNumber].timestamp;
+    }
+
+    function latestRound() external view override returns (uint256) {
+        require(_latestRoundNumber != 0, "No data present");
+
+        return _latestRoundNumber;
+    }
+
+    function latestRoundData()
+        external
+        view
+        override
+        returns (uint80, int256, uint256, uint256, uint80)
+    {
+        require(_latestRoundNumber != 0, "No data present");
+
+        return (
+            uint80(_latestRoundNumber),
+            _roundData[_latestRoundNumber].answer,
+            _roundData[_latestRoundNumber].timestamp,
+            _roundData[_latestRoundNumber].timestamp,
+            uint80(_latestRoundNumber)
+        );
+    }
+
+    function getAnswer(
+        uint256 roundId
+    ) external view override returns (int256) {
+        require(roundId <= _latestRoundNumber, "No data present");
```

Comment:
I guess this and all the rest should also revert if roundId == 0 but probably not so important.

Reply:
Yes, I believe we can live with that (we will be publishing oracles only once there is some data)